File: py/search/linked_list/BinarySearch.py
#!/usr/bin/env python
# -*- coding:gbk -*-

class ListNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class BinarySearch:
    def search(self, head: ListNode, target: int) -> ListNode:
        return self.__search(head.next, None, target)

    def __search(self, start: ListNode, end: ListNode, target: int) -> ListNode:
        if start == end:
            return start if start and target == start.value else None

        middle = self.__getMiddle(start, end)
        if target == middle.value:
            return middle

        if target < middle.value:
            return self.__search(start, middle, target)
        return self.__search(middle.next, None, target)

    def __getMiddle(self, start: ListNode, end: ListNode) -> ListNode:
        slow, fast = start, start.next
        while fast != end and fast.next != end:
            slow, fast = slow.next, fast.next.next
        return slow

File: java/search/linked_list/BinarySearch.java
package search.linked_list;

public class BinarySearch {
    public ListNode search(ListNode head, int target) {
        return search(head.next, null, target);
    }

    private ListNode search(ListNode start, ListNode end, int target) {
        if (start == end) {
            return (null != start && target == start.value) ? start : null;
        }

        ListNode middle = getMiddle(start, end);
        if (target == middle.value) {
            return middle;
        }

        if (target < middle.value) {
            return search(start, middle, target);
        }
        return search(middle.next, null, target);
    }

    private ListNode getMiddle(ListNode start, ListNode end) {
        ListNode slow = start;
        ListNode fast = start.next;
        while (fast != end && fast.next != end) {
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    public class ListNode {
        public int value = 0;
        public ListNode next = null;

        public ListNode(int value) {
            this.value = value;
        }
    }
}
